### PR TITLE
livepatch: do not autoenable livepatch on trusty

### PR DESCRIPTION
**Pull request by blackboxsw**
Due to the memory cost and the cost of also installing
systemd on trusty, livepatch is not auto-enabled.

To obtain series-specific behavior, ua-contracts now provides
a new 'series' key under each entitlement which will contain
any overrides for behavior on a specific series.

UA client now honors overrides via a new util function
apply_series_overrides.

Fixes: #618

### CODE REVIEW

File: uaclient/util.py [hunk 1]
```diff
@@ -358,6 +359,34 @@ def get_platform_info() -> 'Dict[str, str]':
...
+    if not all([isinstance(orig_access, dict), 'entitlement' in orig_access]):
+        raise RuntimeError(
+            'Expected entitlement access dict. Missing "entitlement" key: %s'
+            % orig_access)
+    if copy_dict:
+        updated_access = copy.deepcopy(orig_access)
+    else:
+        updated_access = orig_access
+    series_name = get_platform_info()['series']
+    orig_entitlement = orig_access.get('entitlement', {})
```
Comment (OddBloke) [inline uaclient/util.py]: If we move this up, we can avoid needless `deepcopy` and `get_platform_info` calls in the case we don't have any overrides:

```py
orig_entitlement = orig_access.get('entitlement', {})
overrides = orig_entitlement.get('series', {})
if not overrides:
    return orig_access
```

File: uaclient/tests/test_config.py [hunk 2]
```diff
@@ -18,6 +18,45 @@
...
+        cfg.write_cache('machine-token', token)
+        expected = {
+            'entitlement1': {
+                'entitlement': {'entitled': True, 'type': 'entitlement1'}},
+            'entitlement2': {
+                'entitlement': {'entitled': True, 'type': 'entitlement2'}}}
+        assert expected == cfg.entitlements
+
+    def test_entitlements_use_machine_access_when_present(self, tmpdir):
+        """Return specific machine-access info if present."""
```
Comment (OddBloke) [inline uaclient/tests/test_config.py]: This is orthogonal to this PR, but for my edification: when would we expect machine-access and machine-token to be out of sync?

File: uaclient/entitlements/base.py [hunk 3]
```diff
@@ -225,6 +225,10 @@ def process_contract_deltas(
         delta_entitlement = deltas.get('entitlement', {})
         transition_to_unentitled = bool(delta_entitlement == util.DROPPED_KEY)
         if not transition_to_unentitled:
+            if delta_entitlement:
+                delta_overrides = util.apply_series_overrides(
+                    deltas, copy_dict=False)
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/entitlements/base.py]: I should have mentioned context on this change. I would have preferred to always copy the dict because I don't to rely on internal side-effects that mutate params passed to a function. The issue is that util.DROPPED_KEY is an object() that we create when processing contracts deltas. If we copy when processing deltas, the copied values which used to be util.DROPPED_KEY are now !=  util.DROPPED_KEY so delta processing logic which tries to match util.DROPPED_KEY doesn't match anymore.  I couldn't think of another approach (other than making util.DROPPED_KEY a known-unique-string) which would allow us to copy the delta dictionary and still match util.DROPPED_KEY. We could just make this function always mutate, as any of the callsites can/should cope with the mutated dict just fine.

File: uaclient/util.py (hunk 1 above)
Comment (blackboxsw) [reply uaclient/util.py]: good suggestion will take it.

File: uaclient/tests/test_config.py (hunk 2 above)
Comment (blackboxsw) [reply uaclient/tests/test_config.py]: machine-access *should* always be present for every enabled entitlement. It's a bug if we don't have a machine-access-<service>,json for every entitled service. The only drift could have been introduced if contract entitlements change from entitled to not-entitled. But, in that case we clean up machine-access files as deltas are processed.

File: uaclient/util.py [hunk 4]
```diff
@@ -358,6 +359,34 @@ def get_platform_info() -> 'Dict[str, str]':
...
+    overrides = orig_entitlement.get('series', {}).get(series_name, {})
+    for key, value in overrides.items():
+        try:
+            updated_access['entitlement'][key].update(value)
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/util.py]: series overrides are not full overrides for every value in the top-level directives/obligations or affordances dicts. The only values present in the overrides will be ones that differ from the top-level directives. As such, we need the update to make sure our overrides supplement top-level values.

File: uaclient/entitlements/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply uaclient/entitlements/base.py]: An alternative would be to subvert deepcopy specifically for our sentinel value:

```py
class Sentinel:
    def __deepcopy__(self, _memo):
        return self
```

Note that the memory addresses are the same:

```py
>>> DROPPED_KEY = Sentinel()
>>> {'a': DROPPED_KEY}                                                                                                                                                                                         
{'a': <__main__.Sentinel at 0x7f823056f2b0>}
>>> copy.deepcopy({'a': DROPPED_KEY})                                                                                                                                                                          
{'a': <__main__.Sentinel at 0x7f823056f2b0>}
>>> copy.deepcopy({'a': DROPPED_KEY}) == {'a': DROPPED_KEY}
True
```

File: uaclient/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (OddBloke) [reply uaclient/util.py]: Aha, OK, now I've looked at some actual data this makes sense. A reduced example:

```json
{
    "entitlement": {
        "directives": {
            "aptKey": "13CB4B13",
            "suites": [
                "precise",
                "trusty-security",
                "trusty-updates"
            ]
        },
        "series": {
            "trusty": {
                "directives": {
                    "suites": [
                        "trusty-security",
                        "trusty-updates"
                    ]
                }
            }
        },
        "type": "esm"
    }
}
```

results in

```json
{
    "entitlement": {
        "directives": {
            "aptKey": "13CB4B13",
            "suites": [
                "trusty-security",
                "trusty-updates"
            ]
        },
        "type": "esm"
    }
}
```
after a merge (except with the `series` key included too).

I'm struggling to find the best way to express this, but are we sure that there won't be any further nesting of series overrides below the second level that we're dealing with?

For example, might we ever be given:

```json
{
    "entitlement": {
        "directives": {
            "aptKey": "13CB4B13",
            "nestedDirective": {
                "key1": "value1"
            }
        },
        "series": {
            "trusty": {
                "directives": {
                    "nestedDirective": {
                        "key2": "value2"
                    }
                }
            }
        },
        "type": "esm"
    }
}
```
with the expectation that we will end up with this:
```json
{
    "entitlement": {
        "directives": {
            "aptKey": "13CB4B13",
            "nestedDirective": {
                "key1": "value1",
                "key2": "value2"
            }
        },
        "type": "esm"
    }
}
```

At the moment, I believe we would produce `{"nestedDirective": {"key2": "value2"}}`, dropping the `key1` value.

Or, to put it another way, should we be doing a recursive merge similar to the deltas?

[the remark this answers is not in the document]
Comment (OddBloke) [reply uaclient/util.py]: (To clarify, I'm asking to confirm what the specification on the server side is; if they won't produce this, there is no need for us to handle it.)

File: uaclient/tests/test_config.py (hunk 2 above)
Comment (OddBloke) [reply uaclient/tests/test_config.py]: Thanks!

File: uaclient/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/util.py]: I confirmed with Martin that there are no immediate plans to define object type values under directives, affordances or obligations. And they are aware that ua client will override the entire object value if ua-contracts grows this feature.  Also, if ua-contracts is adding a new directive that has an object value, ua client will need to accept a PR to support said feature and if we have to address specific shortcomings of this approach, we can do it then.

File: uaclient/entitlements/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (blackboxsw) [reply uaclient/entitlements/base.py]: As you said in other comments, all of our callsites are using this object for a fairly short period of time, so mutating it won't be likely to cause us significant issues becase we have a lack of complexity for referencing this object. This is a good alternative suggestion if we were persisting and referencing the object across many callsites in the future. Let's avoid the added work of performing the copy for this. Thanks for the thoughts here!

File: uaclient/util.py (hunk 1 above)
Comment (blackboxsw) [reply uaclient/util.py]: Since we are not going to perform the copy, there is a lot less complexity/cost in this function now. I'll drop this suggestion because our for loop on overrides.items() effectively does no work and doesn't cost much.

File: uaclient/util.py [hunk 5]
```diff
@@ -358,6 +358,25 @@ def get_platform_info() -> 'Dict[str, str]':
     return platform_info
 
 
+def apply_series_overrides(orig_access: 'Dict[str, Any]') -> None:
+    """Apply series-specific overrides to an entitlement dict.
```
Comment (OddBloke) [inline uaclient/util.py]: I'd like the docstring to be explicit about mutation (and probably call out the removal of the `series` key too).
